core/devices/machine_topology.cpp

```diff
@@ -0,0 +1,349 @@
+        std::lock_guard<std::mutex> guard(detail::machine_topology_mutex);
+        if (!detail::machine_topology) {
+            detail::machine_topology = MachineTopology::create();
+            detail::initialized_machine_topology.store(true);
+        }
+    }
+    assert(detail::machine_topology.get() != nullptr);
+    return detail::machine_topology.get();
+}
+
+
+const MachineTopology::io_obj_info *MachineTopology::get_pci_device(
+    const std::string &pci_bus_id) const
+{
+    for (auto id = 0; id < this->pci_devices_.size(); ++id) {
+        if (this->pci_devices_[id].pci_bus_id.compare(0, 12, pci_bus_id, 0,
+                                                      12) == 0) {
+            return &this->pci_devices_[id];
+        }
+    }
+    return nullptr;
+}
+
+
+MachineTopology::MachineTopology()
+{
+#if GKO_HAVE_HWLOC
+
+    // Initialize the topology from hwloc
+    this->topo_ =
+        hwloc_manager<hwloc_topology>(init_topology(), hwloc_topology_destroy);
+    // load objects of type Package . See HWLOC_OBJ_PACKAGE for more details.
+    load_objects(HWLOC_OBJ_PACKAGE, this->packages_);
+    // load objects of type NUMA Node. See HWLOC_OBJ_NUMANODE for more details.
+    load_objects(HWLOC_OBJ_NUMANODE, this->numa_nodes_);
+    // load objects of type Core. See HWLOC_OBJ_CORE for more details.
+    load_objects(HWLOC_OBJ_CORE, this->cores_);
+    // load objects of type processing unit(PU). See HWLOC_OBJ_PU for more
+    // details.
+    load_objects(HWLOC_OBJ_PU, this->pus_);
+    // load objects of type PCI Devices See HWLOC_OBJ_PCI_DEVICE for more
+    // details.
+    load_objects(HWLOC_OBJ_PCI_DEVICE, this->pci_devices_);
+    num_numas_ = hwloc_get_nbobjs_by_type(this->topo_.get(), HWLOC_OBJ_PACKAGE);
+
+#else
+
+    this->topo_ = hwloc_manager<hwloc_topology>();
+
+#endif
+}
+
+
+void MachineTopology::hwloc_binding_helper(
+    const std::tuple<std::vector<MachineTopology::normal_obj_info>,
+                     MachineTopology::hwloc_manager<hwloc_bitmap_s>> &obj,
+    const int *id, const size_type num_ids,
+    const MachineTopology::BitMapType &bitmap_type) const
+{
+#if GKO_HAVE_HWLOC
+    auto bitmap_cur = hwloc_bitmap_alloc();
+    auto bitmap_toset = hwloc_bitmap_alloc();
+    auto bitmap_res = hwloc_bitmap_alloc();
+    // Get current bitmap
+    hwloc_get_cpubind(this->topo_.get(), bitmap_cur, 0);
+    // Set the given ids to a new bitmap
+    for (auto i = 0; i < num_ids; ++i) {
+        GKO_ASSERT(id[i] < std::get<0>(obj).size());
+        GKO_ASSERT(id[i] >= 0);
+        hwloc_bitmap_set(bitmap_toset, std::get<0>(obj)[id[i]].os_id);
+        // Check if this id has been bound before, Binding multiple threads to
+        // same object may not be suitable.
+        if (hwloc_bitmap_isset(std::get<1>(obj).get(),
+                               std::get<0>(obj)[id[i]].os_id)) {
+            hwloc_bitmap_clr(bitmap_toset, std::get<0>(obj)[id[i]].os_id);
+        }
+    }
```

Comment:
I think you are overdoing it with this here. I would remove the `bitmap` thing altogether. Although it's a good idea, this should probably be done one level up and not here. When asking to bind to a core, I would expect it to actually bind to this core. If it's overcrowded, too bad for me.

In general, this way of doing it has flaws as well:
1. You do this per type, but a CUDA host thread could be bound to the same core as a CPU processing executor.
2. In MPI mode, which is the main mode where these aspects need to be taken into account, this will not work anyway.
3. Again in MPI, being able to bind e.g., a CUDA GPU to multiple cores would allow to take care of the oversubscription issue without overcomplicating things: the OS will migrate the threads so that they don't step on each other.
4. I think this would fit better into the automatic `GenericDeviceExecutor` or other executor than here.

Reply:
Sorry I didnt get this part 

> Again in MPI, being able to bind e.g., a CUDA GPU to multiple cores would allow to take care of the oversubscription issue without overcomplicating things: the OS will migrate the threads so that they don't step on each other.

This way of doing it allows you to bind to multiple cores, right ? Or did you mean something else ? 

But otherwise, I do get your point that this is bit over the top